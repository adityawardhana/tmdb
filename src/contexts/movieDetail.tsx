"use client";
import React, { createContext, FC, useContext, useState } from "react";
interface Props {
  children: React.ReactNode;
  initialData: {
    detail?: any;
    images?: any;
    videos?: any;
    credits?: any;
  }; // Replace with your actual movie ID type
}
interface IMovieDetailContext {
  detail?: any;
  images?: any;
  videos?: any;
  credits?: any;
}
const MovieDetailContext = createContext<IMovieDetailContext>({});
const MovieDetailProvider: FC<Props> = ({ children, initialData }) => {
  const [data] = useState(initialData);
  return (
    <MovieDetailContext.Provider value={{ ...data }}>
      {children}
    </MovieDetailContext.Provider>
  );
};

export const useMovieDetailContext = () => useContext(MovieDetailContext);
export default MovieDetailProvider;
