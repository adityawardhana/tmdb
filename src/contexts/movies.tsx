"use client";
import useFetch from "@/hooks/useFetch";
import { moviesServices } from "@/services";
import {
  TFilterKey,
  IFilterMoviesParams,
  GenericResponseList,
  IMovie,
} from "@/types";
import { usePathname, useRouter } from "next/navigation";
import React, { createContext, FC, useContext } from "react";

interface Props {
  initialData?: GenericResponseList<IMovie>; // Replace with your actual data type
  children: React.ReactNode;
  params: IFilterMoviesParams;
}
interface IMoviesContext {
  onParamsChange: (params: IFilterMoviesParams) => void;
  movies?: GenericResponseList<IMovie>;
  filter: TFilterKey;
  page: number;
}
const MoviesContext = createContext<IMoviesContext>({
  onParamsChange: () => {},
  filter: "",
  page: 1,
});

const MoviesProvider: FC<Props> = ({ initialData, params, children }) => {
  const router = useRouter();
  const pathname = usePathname();
  const { filter = "", page = 1 } = params;
  const { data } = useFetch(() => moviesServices.getMovies(params), {
    initialData,
  });

  const onParamsChange = ({ page, filter }: IFilterMoviesParams) => {
    const params = new URLSearchParams();
    params.set("page", (page || 1).toString());
    params.set("filter", filter);
    router.push(`${pathname}?${params.toString()}`);
  };

  return (
    <MoviesContext.Provider
      value={{
        onParamsChange,
        movies: data,
        filter,
        page: Number(page || 1),
      }}
    >
      <main className="flex flex-col gap-8 row-start-2 items-start">
        {children}
      </main>
    </MoviesContext.Provider>
  );
};

export const useMoviesContext = () => useContext(MoviesContext);

export default MoviesProvider;
