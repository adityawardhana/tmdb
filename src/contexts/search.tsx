import useClickOutside from "@/hooks/useClickOutside";
import useFetch from "@/hooks/useFetch";
import { moviesServices } from "@/services";
import { IMovie } from "@/types";
import React, {
  createContext,
  FC,
  useContext,
  useEffect,
  useRef,
  useState,
} from "react";

interface Props {
  children: React.ReactNode;
}

interface ISearchContext {
  query: string;
  loading: boolean;
  results: IMovie[];
  setQuery: React.Dispatch<React.SetStateAction<string>>;
  onClose: () => void;
}

const SearchContext = createContext<ISearchContext>({
  query: "",
  loading: false,
  results: [],
  setQuery: () => {},
  onClose: () => {},
});

const SearchProvider: FC<Props> = ({ children }) => {
  const searchModalRef = useRef(null);
  const [query, setQuery] = useState("");
  const {
    data: { results = [] } = {},
    loading,
    fetch,
  } = useFetch(() => moviesServices.searchMovies(query), { immediate: false });

  const onClose = () => {
    setQuery("");
  };

  useEffect(() => {
    if (!query.length) {
      onClose();
      return;
    }

    const debounceTimeout = setTimeout(() => {
      fetch();
    }, 300);
    return () => clearTimeout(debounceTimeout);
  }, [query]);

  useClickOutside(searchModalRef, () => setQuery(""));

  return (
    <SearchContext.Provider
      value={{ query, loading, results, setQuery, onClose }}
    >
      <header ref={searchModalRef}>{children}</header>
    </SearchContext.Provider>
  );
};

export const useSearchContext = () => useContext(SearchContext);

export default SearchProvider;
