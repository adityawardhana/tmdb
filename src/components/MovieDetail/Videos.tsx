"use client";
import { useMovieDetailContext } from "@/contexts/movieDetail";
import React from "react";

const Videos = () => {
  const { videos = [] } = useMovieDetailContext();

  if (!videos?.results?.length) {
    return <p className="text-sm">No video available.</p>;
  }

  return (
    <section className="flex flex-col p-2 md:p-8" id="movie-videos">
      <p className="text-lg">Videos</p>
      <div className="py-8 gap-2 w-full grid grid-cols-2 md:grid-cols-3">
        {videos.results.map((item: { key: string; name: string }) => (
          <div
            key={item.key}
            className="flex-shrink-0 w-full h-auto md:h-[300px]"
          >
            <iframe
              width="100%"
              height="100%"
              src={`https://www.youtube.com/embed/${item.key}`}
              title={item.name}
              allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
              allowFullScreen
            />
          </div>
        ))}
      </div>
    </section>
  );
};

export default Videos;
