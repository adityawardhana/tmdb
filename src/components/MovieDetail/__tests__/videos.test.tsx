import { MovieDetailContext } from "@/contexts/movieDetail";
import { render, screen } from "@testing-library/react";
import Videos from "../Videos";


it("render Videos component when videos not found", async () => {
  const { container } = render(<Videos />, {
    wrapper: (props) => {
      return (
        <MovieDetailContext.Provider value={{ videos: { results: [] } }}>
          {props.children}
        </MovieDetailContext.Provider>
      );
    },
  });
  expect(container).toMatchSnapshot();
});
