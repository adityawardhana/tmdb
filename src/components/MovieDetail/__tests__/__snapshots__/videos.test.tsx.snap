// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Videos component when videos not found 1`] = `
<div>
  <p
    class="text-sm"
  >
    No video available.
  </p>
</div>
`;
